src/GitHub.App/Models/RepositoryHosts.cs

```diff
@@ -45,78 +40,20 @@ public class RepositoryHosts : ReactiveObject, IRepositoryHosts
-                .Skip(1)  // The first value will be null or something already in the db
-                .SelectMany(enterpriseHost =>
-                {
-                    if (!enterpriseHost.IsLoggedIn)
-                    {
-                        return sharedCache.UserAccount
-                            .InvalidateObject<Uri>(EnterpriseHostApiBaseUriCacheKey)
-                            .Catch<Unit, Exception>(ex =>
-                            {
-                                log.Warn("Failed to invalidate enterprise host uri", ex);
-                                return Observable.Return(Unit.Default);
-                            });
-                    }
-
-                    return sharedCache.UserAccount
-                        .InsertObject(EnterpriseHostApiBaseUriCacheKey, enterpriseHost.Address.ApiUri)
-                        .Catch<Unit, Exception>(ex =>
-                        {
-                            log.Warn("Failed to persist enterprise host uri", ex);
-                            return Observable.Return(Unit.Default);
-                        });
-                });
-
             isLoggedInToAnyHost = this.WhenAny(
                 x => x.GitHubHost.IsLoggedIn,
                 x => x.EnterpriseHost.IsLoggedIn,
                 (githubLoggedIn, enterpriseLoggedIn) => githubLoggedIn.Value || enterpriseLoggedIn.Value)
                 .ToProperty(this, x => x.IsLoggedInToAnyHost);
 
-            // This part is strictly to support having the IConnectionManager request that a connection
-            // be logged in. It doesn't know about hosts or load anything reactive, so it gets
-            // informed of logins by an observable returned by the event
-            connectionManager.DoLogin += RunLoginHandler;
-
-            // monitor the list of connections so we can log out hosts when connections are removed
-            disposables.Add(
-                connectionManager.Connections.CreateDerivedCollection(x => x)
-                .ItemsRemoved
-                .SelectMany(async x =>
-                {
-                    var host = LookupHost(x.HostAddress);
-                    if (host.Address != x.HostAddress)
-                        host = await RepositoryHostFactory.Create(x.HostAddress);
-                    return host;
-                })
-                .Select(h => LogOut(h))
-                .Merge().ToList().Select(_ => Unit.Default).Subscribe());
-
+            connectionManager.ConnectionCreated = ConnectionAdded;
 
-            // Wait until we've loaded (or failed to load) an enterprise uri from the db and then
-            // start tracking changes to the EnterpriseHost property and persist every change to the db
-            disposables.Add(persistEntepriseHostObs.Subscribe());
+            loaded = new TaskCompletionSource<object>();
+            disposables.Add(connectionManager.Connections.ForEachItem(_ => { }, ConnectionRemoved, ConnectionsReset));
```

Comment:
Here we'd ideally just call `ConnectionAdded` as the first callback in `ForEachItem` which would also handle initialization, but because of the hack mentioned above we have to initialize explicitly and use `ConnectionCreated` to create the `RepositoryHost`.